### DISCUSSION TITLE
Kratos serve can't find the configuration file while using Docker Image

### DISCUSSION THREAD

**Discussion by ghost** [ory/kratos#1600] [Q&A] [answered]
# Setup #

As per [Ory Kratos Docker Documentation][1] I run:

```
$ docker pull oryd/kratos:v0.7.1-alpha.1
$ docker run --rm -it oryd/kratos version
Version:			v0.7.1-alpha.1
Build Commit:	4fe76af1302d45ddf4cf3c2c5949311c9cf1f8b8
Build Timestamp:	2021-07-22T17:41:40Z
```

## Running the image in a container ##
What happens here is that no configuration file is specified, so it just errors out the keys that are required.

```
$ docker run oryd/kratos:v0.7.1-alpha.1

The configuration contains values or keys which are invalid:
identity: <nil>
          ^-- one or more required properties are missing

The configuration contains values or keys which are invalid:
selfservice.default_browser_return_url: <nil>
                                        ^-- one or more required properties are missing

The configuration contains values or keys which are invalid:
courier.smtp.connection_uri: <nil>
                             ^-- one or more required properties are missing

time=2021-07-27T17:46:47Z level=fatal msg=Unable to instantiate configuration....
```

# Issue #

When using the Docker Images, Kratos does not recognize a configuration file with the --config flag.

Since containers are ran independently, I figured I'd have to use a file on the Daemon while running the serve command from the daemon and it seems Ory Kratos has a section for this also at [Ory Kratos Docker Image][2]) 

```
docker run --rm -it oryd/kratos serve --config /home/ory/kratos.yml

FATA[2021-07-27T18:35:41Z] Unable to instantiate configuration.          audience=application error=map[message:open /home/ory/kratos.yml: no such file or directory] service_name=Ory Kratos service_version=v0.7.1-alpha.1
```

Relevant Files:

The [configuration][3]

  [1]: https://www.ory.sh/kratos/docs/install#docker
  [2]: https://www.ory.sh/kratos/docs/guides/docker/#creating-custom-docker-image
  [3]: https://github.com/ory/kratos/blob/master/contrib/quickstart/kratos/email-password/kratos.yml

**Comment by aeneasr** [comment, marked as answer]
The problem is that you are not using Docker correctly. Docker does not have access to your host's file system. It's something that runs an isolated container. Since you have not mounted any files to your Docker Image, it can not find the file, because the file, well, doesn't exist.

So even if the file exists locally on your machine:

```
$ ls /home/ory/kratos.yml
```

it doesn't matter to Docker as that's not the way Docker works. You can find out if the file was mounted correctly using:

```
$ docker run -it --entrypoint /bin/ls oryd/kratos:v0.7.1-alpha.1 -la /home/ory
total 8
drwxr-sr-x    2 ory      ory           4096 Jul 22 18:01 .
drwxr-xr-x    1 root     root          4096 Jul 22 18:01 ..
```

As you can see, without any volume mounts, this file doesn't exist.

Please read about docker volume mounting in the Docker documentation: https://docs.docker.com/storage/bind-mounts/

**Reply by ghost** [reply]
Honestly this does look like a docker issue so I can't expect Ory Kratos github to help me figure that out.

Though I will ask:

What is the suggested way to deploy the Ory Kratos clone to a remote server?  Are there any guides on this or really anything at all on remote deploy?

**Reply by vinckr** [reply]
We have some resources in the documentation you might have missed, but if it is mainly a docker issue, it is probably not covered extensively.
There is also a guide on Ory Kratos for native apps: https://www.ory.sh/login-react-native-authentication-example-api/
Another guide for knative & Ory Kratos: https://www.ory.sh/kratos-knative-demo/
We also offer a managed[ Ory Cloud](https://console.ory.sh/) if you want to skip the whole k8s setup 😉

**Comment by fluidum** [comment]
Regarding to https://www.ory.sh/docs/kratos/quickstart#clone-ory-kratos-and-run-it-in-docker
I was investigating similar issue "kratos.yml: no such file or directory" when running **rootless** podman 4.6.2, Ubuntu 22.04
When I went into shell of _kratos_ I got permission denied errors for /etc/config/kratos/ mount binding.

As a fix I added following lines into service blocks of _kratos-migrate_ and _kratos_ into **quickstart.yml**:

```
    environment:
      - USER_ID=1010
      - GROUP_ID=100
    user: "${USER_ID}:${GROUP_ID}"
```

Note: values for environment keys USER_ID and GROUP_ID come by running `id` command. This command is required to run under the user account you plan to execute with.

**Reply by vinckr** [reply]
Thanks for sharing fluidum!
